python/lsst/ts/envsensors/socket_server.py

```
@@ -55,109 +57,106 @@ def __init__(self, port, simulation_mode=0):
         """
         self.log.info("Start called")
-        self._started = True
-        self._server = await asyncio.start_server(
-            self.read_loop, host="127.0.0.1", port=self.port
+        self._server = tcpip.OneClientServer(
+            host=self.host,
+            port=self.port,
+            name=self.log.name,
+            log=self.log,
+            connect_callback=self.connected_callback,
         )
+        await self._server.start_task
 
         # Request the assigned port from the server so the code starting the
         # mock controller can use it to connect.
         if self.port == 0:
-            self.port = self._server.sockets[0].getsockname()[1]
+            self.port = self._server.port
 
         if keep_running:
-            await self._server.serve_forever()
-
-    async def write(self, data):
+            await self._server.server.serve_forever()
+
+    def connected_callback(self, server: tcpip.OneClientServer) -> None:
+        """A client has connected or disconnected."""
+        if not server.connected:
+            self.log.info("Client disconnected.")
+        else:
+            self.log.info("Client connected.")
+        self.read_loop_task.cancel()
+        if server.connected:
+            self.log.info("Starting a new read loop.")
+            self.read_loop_task = asyncio.create_task(self.read_loop())
+
+    async def write(self, data: dict) -> None:
         """Write the data appended with a newline character.
 
         The data are encoded via JSON and then passed on to the StreamWriter
         associated with the socket.
 
         Parameters
         ----------
-        data:
+        data: `dict`
             The data to write.
         """
         self.log.debug(f"Writing data {data}")
         st = json.dumps({**data})
         self.log.debug(st)
-        if self._writer:
-            self._writer.write(st.encode() + b"\r\n")
-            await self._writer.drain()
+        if self._server.writer:
```

I suggest the test be `if self._server.connected` and that you raise an exception if it is not connected.

At present your code silently does nothing if the server is not connected. Silent failures are confusing!

---

I removed the check since it not only is confusing but also not necessary.